CONTRIBUTING.md

```diff
@@ -1,111 +1,190 @@
+   that is not a getter/setter should be documented.
+   * For functions with large bodies, a good practice is to group blocks of
+   related code, and precede them with a blank line and a high-level comment
+   on what the block does.
+
+5. **Benchmarks**: Add micro-benchmarks to support your claims.
+   * As needed, add micro-benchmark to objectively evaluate performance and
+   efficiency trade-offs.
+
+6. **APIs**: Carefully design APIs.
+   * As a library, Velox APIs should be intentional. External API should only 
+   be deliberately created. 
+   * As a rule of thumb, components should be deep and encapsulate as much
+   complexity as possible, and APIs should be narrow, minimizing dependencies
+   across components and preventing implementation details from leaking through
+   the API.
+
+## Adding Scalar Functions
+
+Adding Presto and Spark functions are a good way to get yourself familiar with
+Velox and the coding review process. In addition to the general contribution
+guidelines presented above, here are specific guidelines for contributing
+functions:
 
 1. Read [How to add a scalar function?](https://facebookincubator.github.io/velox/develop/scalar-functions.html) guide.
 
-2. Use the following template for the PR title: Add xxx Presto function (replace xxx with the function name.)
+2. Use the following template for the PR title: Add xxx [Presto|Spark] function (replace xxx with the function name).
+   * Ensure the PR description contains a link to the function documentation
+   from Presto or Spark docs. 
+   * Describe the function semantics and edge cases clearly.
 
-3. Add documentation for the new function to an .rst file under velox/docs/functions directory.
+3. Use Presto or Spark to check the function semantics. 
+   * Try different edge cases to check whether the function returns null, or
+   throws, etc. 
+   * Make sure to replicate the exact semantics.
 
-4. Functions in documentation are listed in alphabetical order. Make sure to
-   place the new function so that the order is preserved.
+4. Add tests exercising common inputs, all possible signatures and corner cases. 
+   * Make sure the test cases are concise and easily readable. 
 
-5. Use Presto to check the function semantics. Try different edge cases to see
-   whether the function returns null, or throws, or does something else. Make sure
-   to replicate Presto semantics exactly.
+5. Make sure that obvious inefficiencies are addressed. 
+   * If appropriate, provide micro-benchmarks to support your claims with data. 
 
-6. Add tests.
+4. Add documentation for the new function to an .rst file under velox/docs/functions directory.
+   * Functions in documentation are listed in alphabetical order. Make sure to
+   place the new function so that the order is preserved.
 
-7. Run [Fuzzer](https://facebookincubator.github.io/velox/develop/testing/fuzzer.html)
+6. Run [Fuzzer](https://facebookincubator.github.io/velox/develop/testing/fuzzer.html)
```

Comment:
nit: Should we add a section about creating or checking up for fuzzer issues as the fuzzer uncovers problems (also adding a note upload fuzzer artifacts).

Reply:
Is this different from the information we have here?

https://facebookincubator.github.io/velox/develop/testing/fuzzer.html

let's sync up offline and discuss what information would make sense to add here.